refactor: process all streaming events for forward compatibility

fede-kamel opened the pull request:
# refactor: process all streaming events for forward compatibility

## Summary

This PR refactors the streaming implementation to process **all event types** instead of filtering to a predefined set. This improves forward compatibility by ensuring any new event types introduced by the API will be automatically processed without requiring SDK updates.

## Changes Made

**Rebased on latest `next` branch** which includes the `response.close()` optimization from codegen (commit 109b771).

### Key Change: Forward Compatibility

**Replaced hardcoded event filtering with generic event processing:**

#### Before (filtering to predefined events):
```python
if (sse.event == "message_start" or 
    sse.event == "message_delta" or 
    sse.event == "message_stop" or ...):
    # Only processes these 6 specific event types
    yield process_data(...)
```

#### After (processes any event):
```python
if sse.event == "completion":
    yield process_data(...)

if sse.event == "ping":
    continue

if sse.event == "error":
    raise error...

# Process ANY other event for forward compatibility
data = sse.json()
if is_dict(data) and "type" not in data:
    data["type"] = sse.event
yield process_data(...)
```

## Benefits

1. **Forward Compatibility** - New event types automatically supported without SDK changes
   - If the API introduces `content_block_thinking`, `tool_use_start`, etc., they'll automatically work
   - No SDK update required for new event types

2. **Cleaner Code** - Simpler logic without predefined event filtering
   - Removed `MESSAGE_EVENTS` frozenset constant
   - More maintainable and easier to understand

3. **API Alignment** - Matches the SDK design philosophy of being forward compatible

4. **Zero Breaking Changes** - Existing functionality preserved
   - All current events still processed correctly
   - Performance maintained with `response.close()` optimization from codegen

## Backward Compatibility

This change maintains 100% backward compatibility. All existing streaming code continues to work identically, with improved forward compatibility for future API evolution.

## Technical Details

- Applies to both sync (`Stream`) and async (`AsyncStream`) implementations
- Keeps the `response.close()` / `await response.aclose()` optimization from codegen
- All events get proper type field assignment if not present

fede-dash commented:
It's been already more than two months this PR was opened. The intention of the PR is to increase the speed of events in your SDK. Difficult to understand why it is taking so long for a clear improvement in the code quality of your SDK.

cc: @RobertCraigie @anthropics/sdk

karpetrosyan commented on src/anthropic/_streaming.py:
```diff
@@ -80,15 +93,8 @@ def __stream__(self) -> Iterator[_T]:
-                or sse.event == "content_block_delta"
-                or sse.event == "content_block_stop"
-            ):
+            # Single, fast membership test instead of multiple string comparisons
```
I like this, it really speeds things up. Plus, even though in this case we have `O(6) = O(1)` complexity for a **single lookup**, our codegen might have N such events, so it's worth it.

karpetrosyan commented on src/anthropic/_streaming.py:
```diff
@@ -198,15 +206,8 @@ async def __stream__(self) -> AsyncIterator[_T]:
-                or sse.event == "content_block_delta"
-                or sse.event == "content_block_stop"
-            ):
+            # same O(1) lookup for consistency between sync/async versions
```
let's have the same comment for sync/async case

karpetrosyan reviewed (commented):
Sorry for delayed answer, I'll try to review it. Thanks so much for your time and efforts, I'll try to get this merged quickly

karpetrosyan commented on src/anthropic/_streaming.py:
```diff
@@ -113,9 +100,12 @@ def __stream__(self) -> Iterator[_T]:
                 )
 
-        # Ensure the entire stream is consumed
-        for _sse in iterator:
```
Ah sorry, we want to keep this part until I port early exit improvement to codegen

fede-kamel commented:
Fixed! I've added back the stream consumption loop. The code now keeps:

```python
# Ensure the entire stream is consumed
for _sse in iterator:
    pass
```

And the async version:

```python
# Ensure the entire stream is consumed  
async for _sse in iterator:
    pass
```

This will remain until you port the early exit improvement to codegen. Changes pushed in commit 2a3e30b.

fede-kamel commented:
## Rebased on latest `next` branch - conflicts resolved ✅

This PR has been rebased onto the latest `next` branch and conflicts have been resolved.

### Resolution Strategy

**Combined both changes:**
1. ✅ **Kept the `response.close()` optimization** from commit 109b771 (the codegen port)
2. ✅ **Kept the forward compatibility refactor** (this PR's main contribution)

### What This PR Changes

The core change is **removing hardcoded event filtering** in favor of **generic event processing**:

**Before:** Only 6 event types were processed (hardcoded list)
```python
if (sse.event == "message_start" or sse.event == "message_delta" or ...):
    yield process_data(...)
```

**After:** ANY event type is processed (forward compatible)
```python
# Handle specific events explicitly
if sse.event == "completion": ...
if sse.event == "ping": continue
if sse.event == "error": raise error

# Process any other event for forward compatibility
data = sse.json()
if is_dict(data) and "type" not in data:
    data["type"] = sse.event
yield process_data(...)
```

This ensures new event types like `content_block_thinking`, `tool_use_start`, etc. will automatically work without SDK updates.

Ready for review! @karpetrosyan
